File: app/database/Resources/PostResource.php
<?php

namespace App\database\Resources;

class PostResource
{
	public static function collection(array $rows): array
	{
		return array_map([self::class, 'transform'], $rows);
	}

	public static function make(array $post): array
	{
		return self::transform($post);
	}

	public static function transform(array $post): array
	{
		if (empty($post)) {
			return [];
		}
		
		return [
			'id' => $post['id'],
			'title' => $post['title'],
			'slug' => $post['slug'],
			'content' => $post['content'],
			'read_time' => $post['read_time'],
			'featured' => $post['featured'],
			'image_url' => $post['image_url'],
			'available_at' => $post['available_at'],
			'category_id' => $post['category_id'] ? $post['category_id'] : null,
			'category' => $post['cat_id'] ? [
				'id' => $post['cat_id'],
				'title' => $post['cat_title'],
				'slug' => $post['cat_slug']
			] : null
		];
	}
}

File: internal/resource/post_resource.go
package resource

import "cat_backend/internal/models"

func MakePost(post *models.Post) map[string]interface{} {
	if post == nil {
		return map[string]interface{}{}
	}

	var category map[string]interface{}
	if post.Category.ID != 0 {
		category = MakeCategory(&post.Category)
	}

	return map[string]interface{}{
		"id":           post.ID,
		"title":        post.Title,
		"slug":         post.Slug,
		"content":      post.Content,
		"read_time":    post.ReadTime,
		"featured":     post.Featured,
		"image_url":    post.ImageURL,
		"available_at": post.AvailableAt,
		"category_id":  post.CategoryID,
		"category":     category,
	}
}

func MakePosts(posts []models.Post) []map[string]interface{} {
	if posts == nil {
		return []map[string]interface{}{}
	}

	var result []map[string]interface{}
	for _, post := range posts {
		result = append(result, MakePost(&post))
	}

	return result
}
